add ccache doc (#4070)
google/iree: docs/developing_iree/ccache.md

# Using `ccache` to build IREE

[`ccache`](https://ccache.dev/) is a compilation cache. In principle, just prepending compiler invocations with `ccache` is all one needs to enable it, e.g.
```shell
ccache clang foo.c -c -o foo.o
```

> takes care of executing `clang` with these arguments and caches the output file `foo.o`. The next invocation then skips executing `clang` altogether.

Reviewer: Could expand on this introduction:

* typical build time savings on one machine
* comparison vs using CMake on its own

(Personally I have fast enough builds on a threadripper without using `ccache` or Bazel RBE)

Author: done!

## Installing and setting up `ccache`

> `ccache` should be easy to install about anywhere. On Debian-based Linux distributions, do:

Reviewer: nit: adjust phrasing

```suggestion
`ccache` is available on most platforms. On Debian-based Linux distributions,
install with:
```

Author: done

```shell
sudo apt install ccache
```

The one `ccache` setting that you probably need to configure is the maximum cache size. The default `5G` is too small for our purposes. To set the cache max size, do this once:
```shell
ccache --max-size=20G
```

> **Tip:** At the moment (late 2020), most of the code we're building is `third_party/llvm-project` so the fundamental limiting factor to how far we can cache away rebuilds is how often that dependency gets updated. Given how frequently it current is update, I'm finding that `20G` is enough to make the `ccache` size not be the limiting factor.

Reviewer: "it **currently** is **updated**"

## Telling CMake to use `ccache`

This is actually a complex topic, there are two different approaches here as explained in this [article](https://crascit.com/2016/04/09/using-ccache-with-cmake/).

### Suggested approach: pass CMake options

In the initial CMake configuration step, set the `IREE_ENABLE_CCACHE` and `LLVM_CCACHE_BUILD` options, like this:
```shell
cmake -G Ninja \
  -DIREE_ENABLE_CCACHE=ON \
  -DLLVM_CCACHE_BUILD=ON \
  ... other options as usual
```

Notes:

> * In the above command line, I explicitly put the `-G Ninja` flag even though that's not a new flag, we're always using `Ninja`. The reason is that the whole approach that these options are relying on only works on certain generators --- `Ninja` and `Makefile`. So if we ever used different generators, we'd have to switch to the other approach, as explained in the above linked [article](https://crascit.com/2016/04/09/using-ccache-with-cmake/).

Reviewer: We recommend Ninja, but do not require it.

```suggestion
* If you use a generator other than `Ninja` see the
  [article](https://crascit.com/2016/04/09/using-ccache-with-cmake/)
  for other approaches.
```

Author: thanks!

* We do need two separate options `IREE_ENABLE_CCACHE` and `LLVM_CCACHE_BUILD`, because of how CMake leaves it up to each project to define how to control the use of `ccache`, and `llvm-project` is a `third_party/` project in IREE. Note that most of the compilation time is spent in `llvm-project`, so `LLVM_CCACHE_BUILD` is the most important flag here.

### Other approach: wrapper scripts

> This is how `ccache` was traditonally used with `make`. This approach is well explained in the above linked [article](https://crascit.com/2016/04/09/using-ccache-with-cmake/). The main reason why you might need it is if you were using another CMake generator than `Ninja` or `Makefile`. For instance, that article covers the case of using the `XCode` generator.

Reviewer: nit: the same thing is said ~3 times in this doc, and "the above linked article" is said many times. Could condense a bit.

Author: good point. condensed!
